### PR TITLE
Make sure only host admins can refund added funds

### PR DESCRIPTION
**Pull request by SudharakaP**
Resolve https://github.com/opencollective/opencollective/issues/5173

### From the Perspective of Collective Admins

image

### From the Perspective of a Host Admin

image

### CODE REVIEW

File: server/graphql/common/transactions.ts [hunk 1]
```diff
@@ -121,11 +122,21 @@ export const canRefund = async (
...
+
+  const isPayeeRootOrHostAdmin = await remoteUserMeetsOneCondition(req, transaction, [isRoot, isPayeeHostAdmin]);
+  /*
+   * We do not allow added funds to be refunded by collective admins as it can lead
+   * to issues and confusions. The host admins will have no idea that the refund was initiated so
+   * the money will not be refunded.
+   */
+  if (transaction.kind === TransactionKind.ADDED_FUNDS && !isPayeeRootOrHostAdmin) {
+    return false;
+  }
```
Comment (znarf) [inline server/graphql/common/transactions.ts:134]:
```suggestion
  if (transaction.kind === TransactionKind.ADDED_FUNDS) {
    return remoteUserMeetsOneCondition(req, transaction, [isRoot, isPayeeHostAdmin]);
  }
  
```

File: server/graphql/common/transactions.ts [hunk 2]
```diff
@@ -121,11 +122,21 @@ export const canRefund = async (
...
   const timeLimit = moment().subtract(30, 'd');
   const createdAtMoment = moment(transaction.createdAt);
   const transactionIsOlderThanThirtyDays = createdAtMoment < timeLimit;
   if (transactionIsOlderThanThirtyDays) {
-    return remoteUserMeetsOneCondition(req, transaction, [isRoot, isPayeeHostAdmin]);
+    return isPayeeRootOrHostAdmin;
```
Comment (znarf) [inline server/graphql/common/transactions.ts:139]:
```suggestion
    return remoteUserMeetsOneCondition(req, transaction, [isRoot, isPayeeHostAdmin]);
```

File: server/graphql/common/transactions.ts [hunk 3]
```diff
@@ -121,11 +122,21 @@ export const canRefund = async (
   if (transaction.type !== TransactionTypes.CREDIT || transaction.OrderId === null || transaction.isRefund === true) {
     return false;
   }
+
+  const isPayeeRootOrHostAdmin = await remoteUserMeetsOneCondition(req, transaction, [isRoot, isPayeeHostAdmin]);
```
Comment (znarf) [inline server/graphql/common/transactions.ts:126]:
```suggestion
```

**Comment by SudharakaP** [comment]
> Proposed code would have worked but it's making things more complicated than they should.
> 
> Made some proposals.

Thanks for the suggestions. I've changed the code accordingly. 👍🏽

**Review by znarf — CHANGES_REQUESTED** [review]
Why not simply do the following?

```transactionIsOlderThanThirtyDays || transaction.kind === TransactionKind.ADDED_FUNDS``` ?

File: server/graphql/common/transactions.ts [hunk 4]
```diff
@@ -121,10 +122,14 @@ export const canRefund = async (
...
     return false;
   }
+
   const timeLimit = moment().subtract(30, 'd');
   const createdAtMoment = moment(transaction.createdAt);
   const transactionIsOlderThanThirtyDays = createdAtMoment < timeLimit;
-  if (transactionIsOlderThanThirtyDays) {
+  /*
+   * To ensure proper accounting, we only allow added funds to be refunded by Host admins and never by Collective admins.
+   */
```
Comment (znarf) [inline server/graphql/common/transactions.ts:131]: The description is not accurate anymore as it doesn't explain `transactionIsOlderThanThirtyDays`.

**Review by znarf — CHANGES_REQUESTED** [review]
Code comment is now outdated.

File: server/graphql/common/transactions.ts (hunk 4 above)
Comment (SudharakaP) [reply server/graphql/common/transactions.ts]: I've added another line to emphasize that the `transactionIsOlderThanThirtyDays` condition as well. We didn't have a comment about this part before, so I am not exactly sure if there's a reason for this 30 day limitation; if so let me know and I'll update the comment. 🤔

**Comment by SudharakaP** [comment]
> Why not simply do the following?
> 
> `transactionIsOlderThanThirtyDays || transaction.kind === TransactionKind.ADDED_FUNDS` ?

@znarf : Thanks. This is done. I also noticed that we should add some tests for this and I've done that as well. When adding tests I've noticed we use a wrong function to test the transaction rejections which I've opened a new PR; 
https://github.com/opencollective/opencollective-api/pull/7115

**Review by znarf — APPROVED** [review]

**Pull request merged**